## ひろしまQuest2020#stayhome【球種予測部門】
#### all_merge_16
- 球団ごとの年棒ランク
- 選手IDごとの2017年のコース実績

#### all_merge_15
- 左右で球種/コースを別に集計

#### all_merge_14
- 左右で球種を別に集計

#### all_merge_13
- 前の投球からの時間差

In [21]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import lightgbm as lgb
import gc
%matplotlib inline
pd.set_option('display.max_Columns', 100)

In [22]:
ALL_MERGE = 'intermediate/all_merge_16.f'
SUBMIT = 'submit/ball_16.csv'

In [23]:
all_pitch = pd.read_feather(ALL_MERGE)
all_pitch.shape

(778767, 181)

In [24]:
print(all_pitch.shape)
column_cnt = len(all_pitch.columns)
all_pitch.head()

(778767, 181)


No  ball  course  No_in_game  pitch_cnt_in_game  inning  bat_cnt_in_inning  \
0   0   0.0     0.0           1                  1       1                  1   
1   1   0.0     8.0           2                  2       1                  1   
2   2   0.0     5.0           3                  3       1                  1   
3   3   0.0    12.0           4                  4       1                  2   
4   4   0.0     8.0           5                  5       1                  2   

   pitch_cnt_in_bat  pitch_order  player_cnt_in_game  pitcher_cnt_in_game  \
0                 1            1                   1                    1   
1                 2            1                   1                    2   
2                 3            1                   1                    3   
3                 1            1                   2                    4   
4                 2            1                   2                    5   

   pitcher_cnt_in_inning  bat_order  bat_cnt_in_game  home_point  away_point  \
0                      1          1                1           0           0   
1                      2          1                1           0           0   
2                      3          1                1           0           0   
3                      4          2                1           0           0   
4                      5          2                1           0           0   

   out_cnt  ball_cnt  strike_cnt  pitch_LR  bat_LR  bc_straight  bc_curve  \
0        0         0           0       0.0     1.0     0.464734  0.110359   
1        0         0           1       0.0     1.0     0.442351  0.093302   
2        0         0           2       0.0     1.0     0.504212  0.069963   
3        1         0           0       0.0     0.0     0.475961  0.101792   
4        1         0           1       0.0     0.0     0.463496  0.099763   

   bc_slider  bc_shoot   bc_fork  bc_changeup  bc_sinker  bc_cutball  \
0   0.151086  0.097383  0.067492     0.043210   0.012933    0.052803   
1   0.109348  0.078518  0.141711     0.065447   0.017669    0.051654   
2   0.065568  0.046520  0.211722     0.059158   0.019231    0.023626   
3   0.226048  0.066617  0.042131     0.018345   0.003343    0.065762   
4   0.194714  0.067542  0.091198     0.023901   0.004486    0.054898   

   bc_course00  bc_course01  bc_course02  bc_course03  bc_course04  \
0     0.064837     0.095199     0.082009     0.040512     0.063252   
1     0.044893     0.071306     0.068422     0.032002     0.043180   
2     0.027656     0.045421     0.045238     0.025275     0.023810   
3     0.034319     0.043647     0.027090     0.054336     0.065374   
4     0.024390     0.034832     0.025124     0.039481     0.047557   

   bc_course05  bc_course06  bc_course07  bc_course08  bc_course09  \
0     0.053916     0.030534     0.046508     0.038328     0.163762   
1     0.036149     0.029388     0.042910     0.033625     0.152078   
2     0.027106     0.023077     0.032784     0.022711     0.202930   
3     0.045046     0.062653     0.097944     0.085507     0.098410   
4     0.033526     0.049596     0.080512     0.069337     0.085243   

   bc_course10  bc_course11  bc_course12  first  second  third  base_cnt  \
0     0.057728     0.130872     0.132543      0       0      0         0   
1     0.078157     0.203642     0.164248      0       0      0         0   
2     0.081319     0.285165     0.157509      0       0      0         0   
3     0.098527     0.055307     0.231840      0       0      0         0   
4     0.121054     0.076923     0.312424      0       0      0         0   

   top_bot  role  pos_pit  date_from_opening  ...  salary_year_bat  \
0        1     1        0                  0  ...           1875.0   
1        1     1        0                  0  ...           1875.0   
2        1     1        0                  0  ...           1875.0   
3        1     1        0                  0  ...            320.0   
4        1     1        0  

### train
- 行数: 257117

In [25]:
train = all_pitch.dropna(subset=['course'])
train.shape

(257117, 181)

### test
- 行数: 521650

In [26]:
test = all_pitch[all_pitch['course'].isnull()]
test.shape

(521650, 181)

In [27]:
del all_pitch
gc.collect()

66

In [28]:
train_d = train.drop([
    'No', 
    'course', 
    'ball'
], axis=1)

In [29]:
test_d = test.drop([
    'No', 
    'course', 
    'ball'
], axis=1)

## Dataset作成

In [30]:
lgb_train = lgb.Dataset(train_d, train['ball'])

### LGBM学習
- multiclass : クラス数=8
- tuning結果のパラメータを使用
- 'learning_rate' : 0.1

In [31]:
%%time
lgb_param = {
        'objective' : 'multiclass',
        'boosting_type': 'gbdt',
        'metric' : 'multi_logloss',
        'num_class' : 8,
        'seed' : 0,
        'learning_rate' : 0.1,
        'lambda_l1': 6.9923570049658075, 
        'lambda_l2': 0.002378623984798833, 
        'num_leaves': 18, 
        'feature_fraction': 0.45199999999999996, 
        'bagging_fraction': 0.9799724836460725, 
        'bagging_freq': 4, 
        'min_child_samples': 20
    }

# 学習
cv_results = lgb.cv(lgb_param, lgb_train,
                    num_boost_round=15000,
                    early_stopping_rounds=100,
                    verbose_eval=100,
                    nfold=4)

num_boost_round = len(cv_results['multi_logloss-mean'])
print('Best num_boost_round:', num_boost_round)
best_cv_score = cv_results['multi_logloss-mean'][-1]
print('Best CV score:', best_cv_score)

[100]	cv_agg's multi_logloss: 1.14331 + 0.00225468
[200]	cv_agg's multi_logloss: 1.12018 + 0.00226051
[300]	cv_agg's multi_logloss: 1.11287 + 0.00261688
[400]	cv_agg's multi_logloss: 1.10896 + 0.00257113
[500]	cv_agg's multi_logloss: 1.10653 + 0.00270896
[600]	cv_agg's multi_logloss: 1.10478 + 0.00282483
[700]	cv_agg's multi_logloss: 1.10373 + 0.00287215
[800]	cv_agg's multi_logloss: 1.10312 + 0.0030187
[900]	cv_agg's multi_logloss: 1.10285 + 0.00309544
[1000]	cv_agg's multi_logloss: 1.10272 + 0.00306425
Best num_boost_round: 986
Best CV score: 1.1027008141247947
CPU times: user 2h 1min 22s, sys: 2min 19s, total: 2h 3min 41s
Wall time: 30min 57s


In [32]:
best_iter = int(num_boost_round * 1.1)

### 全体で再学習

In [33]:
%%time
lgb_model = lgb.train(lgb_param, lgb_train, num_boost_round=best_iter)

CPU times: user 38min 11s, sys: 44.8 s, total: 38min 55s
Wall time: 9min 42s


### Feature Importance

In [34]:
fi = lgb_model.feature_importance()
fn = lgb_model.feature_name()
df_feature_importance = pd.DataFrame({'feat_name':fn, 'feat_imp':fi})
df_feature_importance.sort_values('feat_imp', inplace=True)
df_feature_importance.tail(30)

feat_name  feat_imp
18             bc_straight      1664
29             bc_course03      1674
93                    fork      1676
6       player_cnt_in_game      1685
155      rank_year_dif_p-b      1697
166         salary_dif_b-c      1698
115             height_bat      1713
170  salary_x_year_dif_b-c      1717
116             weight_bat      1727
165            bmi_dif_p-c      1758
90                   curve      1775
153    salary_year_dif_p-b      1838
171      rank_year_dif_b-c      1917
53              point_diff      1961
132           bat_game_cnt      2008
130                bmi_bat      2008
156    rank_x_year_dif_p-b      2016
169    salary_year_dif_b-c      2148
172    rank_x_year_dif_b-c      2162
91                  slider      2206
131             batter_cnt      2308
1        pitch_cnt_in_game      2326
0               No_in_game      2367
8    pitcher_cnt_in_inning      2461
157            bmi_dif_p-b      2597
173            bmi_dif_b-c      2716
47             elapsed_min      2925
7      pitcher_cnt_in_game      3246
46       date_from_opening      4819
48                min_diff      5722

In [35]:
%%time
predict = lgb_model.predict(test_d, num_iteration = lgb_model.best_iteration)

CPU times: user 7min 27s, sys: 1.17 s, total: 7min 28s
Wall time: 1min 55s


## 結果出力

In [36]:
submit = pd.DataFrame(predict)
submit.reset_index(inplace=True)
print(submit.shape)
submit.head()

(521650, 9)


index         0         1         2         3         4         5  \
0      0  0.466142  0.057530  0.155094  0.262030  0.003081  0.000031   
1      1  0.489377  0.108294  0.135576  0.227138  0.006087  0.000041   
2      2  0.401369  0.121335  0.151008  0.120021  0.133522  0.000119   
3      3  0.387750  0.132434  0.199023  0.228204  0.017727  0.000089   
4      4  0.226666  0.054395  0.336016  0.297710  0.013503  0.000076   

          6         7  
0  0.000037  0.056054  
1  0.000018  0.033469  
2  0.000059  0.072566  
3  0.000031  0.034743  
4  0.000026  0.071607

### 2017の平均
- ストレート           0.468254
- カーブ             0.074725
- スライダー           0.185806
- シュート            0.070633
- フォーク            0.083013
- チェンジアップ         0.054217
- シンカー            0.011361
- カットボール          0.051992

In [37]:
submit.mean()

index    260824.500000
0             0.525531
1             0.073459
2             0.182520
3             0.054714
4             0.070680
5             0.044504
6             0.006312
7             0.042280
dtype: float64

In [38]:
submit.to_csv(SUBMIT, header=False, index=False)

### SIGNATE

In [39]:
'signate submit --competition-id=275 ./{} --note feat={}_cv={}'.format(SUBMIT, column_cnt, best_cv_score)

'signate submit --competition-id=275 ./submit/ball_16_tune.csv --note feat=181_cv=1.1027008141247947'

### GCS

In [40]:
'gsutil cp ./{} gs://yasunobu_baseball/submit/'.format(SUBMIT)

'gsutil cp ./submit/ball_16_tune.csv gs://yasunobu_baseball/submit/'

### 評価結果
#### 16 (181) ... 選手IDごとの2017年のコース実績
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1178
- Best CV score: 1.1021930539837317
- Wall time: 12min 35s
- 評価結果  : 1.3993836

#### 15 (153) ... 左右で球種/コースを別に集計
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1189
- Best CV score: 1.1013869435441306
- Wall time: 11min 9s
- 評価結果  : 未提出

#### 14 (153) ... 左右で球種を別に集計
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1240
- Best CV score: 1.1021952294615331
- Wall time: 11min 34s
- 評価結果  : 1.3981900

#### 13 (157) ... 前の投球からの時間差を追加
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1416
- Best CV score: 1.1062463941695753
- Wall time: 13min 10s
- 評価結果  : 1.4012184

#### 12 (156) ... コースの集計を追加
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1353
- Best CV score: 1.106942086205103
- Wall time: 27min 41s
- 評価結果  : 1.4016077

#### 11 (143) ... ボールカウントの集計を追加
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1209
- Best CV score: 1.1075153721728936
- Wall time: 23min 19s
- 評価結果  : 未提出

#### 10_tune (135) ... 再チューニング
- 'num_leaves' : 15,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1260
- Best CV score: 1.1118897437428692
- Wall time: 8min 40s
- 評価結果  : 1.4060397

#### 10s (135) ... trainを75%サンプリング
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1412
- Best CV score: 1.1190287120710383
- Wall time: 12min 5s
- 評価結果  : 1.4310869

#### 10 (135) ... 5相当に戻した
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1112
- Best CV score: 1.111265000190272
- Wall time: 9min 30s
- 評価結果  : 1.4008712

#### 9 (136) ... 5相当に戻した=>NG
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1106
- Best CV score: 1.1113177029272225
- Wall time: 25min 44s (local)
- 評価結果  : 1.5133521

#### 8 (157)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1100
- Best CV score: 1.1093728644236824
- Wall time: 10min 4s
- 評価結果  : 1.4925832

#### 7 (235)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1204
- Best CV score: 1.108590082437519
- Wall time: 12min 56s
- 評価結果  : 1.4946937

#### 6 (207)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1135
- Best CV score: 1.1086837935965734
- Wall time: 11min 54s
- 評価結果  : 1.3999194

#### 5-tune (参考)
- 'num_leaves' : 18,
- 'learning_rate' : 0.05,
- Best num_boost_round: 2395
- Best CV score: 1.1103399384119979
- Wall time: 18min 29s (n1-standard-8)

#### 5-tune (135)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- CV score: 1.1112402566616455
- 評価結果  : 1.3991780
- CV Wall time: 27min 48s (local) -> 18min 24s (n1-standard-4) -> 9min 3s (n1-standard-8)

#### 5 (135)
- 'num_leaves' : 31,
- 'learning_rate' : 0.1,
- CV score: 1.1161825304353852
- 評価結果  : 1.4252286

#### 4 (128)
- CV score: 1.1158111693845418
- 評価結果  : 1.4299265

#### 3 (97)
- CV score: 1.1172290313076036
- 評価結果  : 1.4335806